### PR TITLE
Shared Resources Overhaul

### PR DESCRIPTION
**Pull request by bert-mccutchen**
## Why are these changes being made?
With Phillips WiZ releasing more devices of different types, it would be nice to support them. This change allows for more device types to be added and supported.

## Changes
* Created a new Shared module that houses most of the core logic.
    * Device Connections
    * Device Network Discovery
    * Device Groups
    * Device State
    * Base Events
 * Most of the core functionality that the Lighting module uses is inherited from the Shared module.
 * Added a Power module to support plugs.
 * Lots and lots of specs.

### CODE REVIEW

**Comment by ppostma** [comment]
Hi! Nice to see these changes!

I've tried it out with my devices, which are two [Filament amber ST64 E27](https://www.wizconnected.com/en/consumer/products/8718699787332/) lights and a [Smart Plug Type F](https://www.wizconnected.com/en/consumer/products/8719514552685/).

```ruby
group = Wizrb::Shared::Discover.all
group.each { |device| puts device.inspect }
```

Gives the output:
```shell
#<Wizrb::Shared::Products::Device:0x00007fd4a921abd0 @ip="192.168.178.80", @port=38899, @state=#<Wizrb::Shared::State:0x00007fd4a921ab80 @state={:state=>true}>, @connection=#<Wizrb::Shared::Connection:0x00007fd4a921ab30 @ip="192.168.178.80", @port=38899, @connection_id="61fae849", @socket=#<UDPSocket:fd 11, AF_INET, 192.168.178.27, 53999>>, @system_config={"mac"=>"a1b2c3d4e5f6", "homeId"=>3983939, "roomId"=>6025536, "rgn"=>"eu", "moduleName"=>"ESP10_SOCKET_06", "fwVersion"=>"1.24.0", "groupId"=>0, "drvConf"=>[20, 2], "ping"=>0}>
#<Wizrb::Shared::Products::Device:0x00007fd4a9218b78 @ip="192.168.178.79", @port=38899, @state=#<Wizrb::Shared::State:0x00007fd4a9218b28 @state={:state=>false}>, @connection=#<Wizrb::Shared::Connection:0x00007fd4a9218ad8 @ip="192.168.178.79", @port=38899, @connection_id="61fae849", @socket=#<UDPSocket:fd 14, AF_INET, 192.168.178.27, 55338>>, @system_config={"mac"=>"a1b2c3d4e5f6", "homeId"=>3983939, "roomId"=>6025536, "rgn"=>"eu", "moduleName"=>"ESP21_SHTW_01", "fwVersion"=>"1.25.0", "groupId"=>0, "ping"=>0}>
#<Wizrb::Shared::Products::Device:0x00007fd4a9222df8 @ip="192.168.178.78", @port=38899, @state=#<Wizrb::Shared::State:0x00007fd4a9222da8 @state={:state=>false}>, @connection=#<Wizrb::Shared::Connection:0x00007fd4a9222d58 @ip="192.168.178.78", @port=38899, @connection_id="61fae849", @socket=#<UDPSocket:fd 15, AF_INET, 192.168.178.27, 55796>>, @system_config={"mac"=>"a1b2c3d4e5f6", "homeId"=>3983939, "roomId"=>6025536, "rgn"=>"eu", "moduleName"=>"ESP21_SHTW_01", "fwVersion"=>"1.25.0", "groupId"=>0, "ping"=>0}>
```

And controlling a SmartPlug works as expected:

```ruby
plug = Wizrb::Lighting::Products::SmartPlug.new(ip: '192.168.178.80')
plug.power_on
```

Also, the brightness is now returned correctly for the lights.

I noticed when using `Wizrb::Lighting::Discover`, the Smart Plug is included in the discovered devices, but as a `DimableLight`. I'm not sure if you want to include the Smart Plug in this list as it isn't technically a light. But you can derive it from the module name in the output I shared above.

Maybe add a few words in the readme about device discovery and the power module? :wink:

**Comment by ppostma** [comment]
Btw, the SmartPlug, shouldn't its namespace be `Wizrb::Power::Products::SmartPlug` ?

**Comment by bert-mccutchen** [comment]
Good findings! Unfortunately module names is one thing i'm having a hard time with since I only have the one type of lights.

**Comment by bert-mccutchen** [comment]
Alright, hopefully it detects your plug properly now @ppostma!

File: lib/wizrb/lighting/discover.rb [hunk 1]
```diff
@@ -1,125 +1,41 @@
...
+        if module_name.include?(Wizrb::Lighting::Products::RgbLight::MODULE_NAME_IDENTIFIER)
           Wizrb::Lighting::Products::RgbLight.new(ip: ip, port: port)
-        else
+        elsif module_name.include?(Wizrb::Lighting::Products::TunableLight::MODULE_NAME_IDENTIFIER)
+          Wizrb::Lighting::Products::TunableLight.new(ip: ip, port: port)
+        elsif module_name.include?(Wizrb::Lighting::Products::DimableLight::MODULE_NAME_IDENTIFIER)
           Wizrb::Lighting::Products::DimableLight.new(ip: ip, port: port)
         end
+
+        nil
```
Comment (ppostma) [inline lib/wizrb/lighting/discover.rb:34]:
```suggestion
```

File: lib/wizrb/power/discover.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      end
+
+      def resolve_device(ip:, port: 38_899)
+        module_name = Wizrb::Shared::Products::Device.new(ip: ip, port: port).module_name
+
+        if module_name.include?(Wizrb::Power::Products::SmartPlug::MODULE_NAME_IDENTIFIER)
+          Wizrb::Power::Products::SmartPlug.new(ip: ip, port: port)
+        end
+
+        nil
```
Comment (ppostma) [inline lib/wizrb/power/discover.rb:28]:
```suggestion
```

**Comment by ppostma** [comment]
Thanks, with the two suggested changes, then the light & plug discovery works OK!

**Review by ppostma — APPROVED** [review]

**Comment by ppostma** [comment]
I think you missed my two suggested changes, without it the resolve method always returns nil and no devices will be returned in the discovery.

**Comment by bert-mccutchen** [comment]
Ahh! GitHub Mobile!!! It doesn't show all messages by default.

**Comment by bert-mccutchen** [comment]
Fixed! I've also added automated releases to GitHub Packages and RubyGems! No more forgetting. 🥳 🎉

**Pull request merged**